enhancements/proposals/1304-external-seeds/README.md

```diff
@@ -0,0 +1,147 @@
+
+### Risks and Mitigations
+
+#### Lookup of seed addresses on ScyllaDB initialization
+Currently, ScyllaDB performs an eager lookup of provided seeds on initialisation: https://github.com/scylladb/scylladb/blob/39ca07c49b257d37856164ea5b845f7b98224b98/init.cc#L35.
+An error in address lookup will hence cause ScyllaDB process to fail initialisation. This is expected during bootstrap.
+
+However, this is also going to cause an already bootstrapped node to fail after it has been restarted, e.g.:
+```
+ERROR 2023-08-03 08:35:54,663 [shard 0] init - Bad configuration: invalid value in 'seeds': 'dummy.scylla.svc.cluster.local': std::system_error (error C-Ares:4, dummy.scylla.svc.cluster.local: Not found)
+```
+
+Although the [documentation](https://opensource.docs.scylladb.com/stable/kb/seed-nodes.html) claims, that "**Once the nodes have joined the cluster, seed node has no function**", introducing the option to control the external seeds might cause instability to the ScyllaDB nodes controlled by the operator due to the above.
+To mitigate this, an issue has been opened with ScyllaDB core: [scylladb/scylladb#14945](https://github.com/scylladb/scylladb/issues/14945).
+
+## Design Details
+
+### ScyllaCluster API
+
+The following extension to the existing ScyllaCluster CR API is proposed:
+```go
+// ScyllaClusterSpec defines the desired state of Cluster.
+type ScyllaClusterSpec struct {
+    ...
+	
+	// externalSeeds specifies the external seeds to propagate to ScyllaDB binary on startup as "seeds".
+	// See https://opensource.docs.scylladb.com/stable/kb/seed-nodes.html.
+	// +optional
+	ExternalSeeds []string `json:"externalSeeds,omitempty"`
+}
+```
+
+The provided `externalSeeds` are passed to the sidecar through a new `--external-seeds` flag. StatefulSet reconciliation in ScyllaDB operator is modified accordingly.
+The sidecar's strategy for choosing seeds to provide to ScyllaDB processes is modified to follow the below pseudocode:
+```go
+func chooseSeeds(externalSeeds []string, seedFromLocalDC string) []string {
+  if is first node in DC {
+    if externalSeeds not empty {
+      return externalSeeds
+    }
+    return seedFromLocalDC
+  }
+
+  return append(externalSeeds, seedFromLocalDC)
+}
+```
+The strategy for getting a seed from local DC (`seedFromLocalDC`) is not modified, i.e. for the first running node in the DC, its own address (`broadcast_address`) is selected, otherwise an address of one other running node from its own DC is selected.
+
+It is important that the external seeds, and only the external seeds, are provided as seeds for the first running node in a datacenter, if they are at all provided by the end user. If external seeds were only appended to the node's own IP, it could result in creating split clusters in case of lack of connectivity between the datacenters.
+For the other nodes, a seed from their own DC is appended to the external seeds for increased resiliency.
```

Comment:
can other nodes default to it's own IP when the rest is unavailable resulting in the same situation as described for the first node?

Reply:
No, as described there the seeds they are provided on bootstrap are external seeds + an address of one other running node from their DC. The first one in the DC won't be able to get up if there is no connectivity between them and external seeds. For the the other nodes in the DC, they will learn the topology from the other node in their DC even if there's no connectivity to the external seeds. So it's impossible that they form a split cluster at this point.

Reply:
Scenario:
- DC A is already bootstrapped
- dc-b-node-0 starts and bootstraps
- dc-b-node-1 pods is created but before it starts, dc-b-node-0 gets evicted
- dc-b-node-1 has no external seeds (as for this proposal) and hits
https://github.com/scylladb/scylla-operator/blob/d8e7d3c7580bbdd613ca673b115137a70c0f9294/pkg/sidecar/identity/member.go#L74-L77
which uses it's own Service IP as a seed instead of the first node or the other DC - witting it down it seems that might be an issue in the single DC case as well

Reply:
> dc-b-node-1 has no external seeds (as for this proposal)

I don't see this situation happening. There's two possibilities for the seeds if external seeds were provided to a DC:
- *only* external seeds for first running node in DC
- external seeds + one seed from own DC for other nodes

So in this situation, dc-b-node-1 would be provided with only the external seeds and should join the cluster correctly. 

As for the single DC, yes, it looks like it might be an issue, but it's not introduced by this proposal.